hawkular-dmr-client/src/main/java/org/hawkular/dmr/api/OperationBuilder.java

```diff
@@ -0,0 +1,446 @@
+    R extends OperationResult<?>> {
+        protected final ModelNode baseNode = new ModelNode();
+
+        public ModelNode build() {
+            return baseNode;
+        }
+
+        protected R createResult(ModelNode request, ModelNode result) {
+            return (R) new OperationResult(request, result);
+        }
+
+        public R execute(ModelControllerClient client) {
+            ModelNode request = build();
+            ModelNode result;
+            try {
+                result = client.execute(request);
+            } catch (IOException e) {
+                throw new DmrApiException(e);
+            }
+            log.tracef("Executed [%s] built by [%s] with result [%s]", request, getClass().getName(), result);
+            return createResult(request, result);
+        }
+
+    }
+
+    public abstract static class AbstractSingleOperationBuilder //
+    <T extends AbstractSingleOperationBuilder<?, R>, R extends OperationResult<?>>
+            extends AbstractOperationBuilder<T, R> {
+        protected final BatchOperationBuilder<BatchOperationBuilder<?>> batch;
+
+        private AbstractSingleOperationBuilder(BatchOperationBuilder<BatchOperationBuilder<?>> batch,
+                String operationName) {
+            super();
+            this.batch = batch;
+            this.baseNode.get(OP).set(operationName);
+        }
+
+        public AddressBuilder<AddressBuilder<?, T>, T> address() {
+            return new AddressBuilder<AddressBuilder<?, T>, T>((T) this);
+        }
+
+        public T address(ModelNode addess) {
+            baseNode.get(ADDRESS).set(addess);
+            return (T) this;
+        }
+
+        public BatchOperationBuilder<?> parentBuilder() {
+            return batch.operation(build());
+        }
+
+    }
+
+    public static class AddOperationBuilder<T extends AddOperationBuilder<?>>
+            extends AbstractSingleOperationBuilder<T, OperationResult<?>> {
+
+        private AddOperationBuilder(BatchOperationBuilder<BatchOperationBuilder<?>> batch) {
+            super(batch, ADD);
+        }
+
+        /**
+         * If {@code value == null}, no entry is added. Otherwise, the attribute will be added.
+         *
+         * @param name
+         * @param value
+         * @return
+         */
+        public T attribute(String name, String value) {
+            if (value != null) {
+                baseNode.get(name).set(value);
+            }
+            return (T) this;
+        }
+
+        public T attribute(String name, boolean value) {
+            baseNode.get(name).set(value);
+            return (T) this;
+        }
+
+        public T valueAttribute(String value) {
+            return attribute(VALUE, value);
+        }
+
+    }
+
+    public static class AddressBuilder<T extends AddressBuilder<?, P>, P extends AbstractSingleOperationBuilder<?, ?>> {
+        private final ModelNode baseNode = new ModelNode();
```

Comment:
curious why you didn't use the Address DMR client class - seems like that could have been useful here